File: src/geegoo_agent/infra/events.py
"""In-process event bus (L0)."""

from __future__ import annotations

import logging
from collections import defaultdict
from collections.abc import Callable
from typing import Any, Protocol

logger = logging.getLogger(__name__)

EventHandler = Callable[[dict[str, Any]], None]


class EventBus(Protocol):
    def emit(self, event: str, payload: dict[str, Any]) -> None: ...
    def on(self, event: str, handler: EventHandler) -> None: ...


class InProcessEventBus:
    """Synchronous in-process event bus; handler errors are logged, not propagated."""

    def __init__(self) -> None:
        self._handlers: dict[str, list[EventHandler]] = defaultdict(list)
        self.history: list[tuple[str, dict[str, Any]]] = []

    def on(self, event: str, handler: EventHandler) -> None:
        self._handlers[event].append(handler)

    def emit(self, event: str, payload: dict[str, Any]) -> None:
        record = dict(payload)
        self.history.append((event, record))
        for handler in self._handlers.get(event, []):
            try:
                handler(record)
            except Exception:
                logger.exception("event_handler_failed", extra={"event": event})

File: internal/infra/events.go
package infra

import "log"

// EventHandler receives one event payload.
type EventHandler func(payload map[string]any)

// EventRecord is one emitted event.
type EventRecord struct {
	Event   string
	Payload map[string]any
}

// EventBus emits in-process events (Python InProcessEventBus parity).
type EventBus struct {
	handlers map[string][]EventHandler
	History  []EventRecord
}

// NewEventBus creates an empty bus.
func NewEventBus() *EventBus {
	return &EventBus{handlers: make(map[string][]EventHandler)}
}

// On registers a handler for an event name.
func (b *EventBus) On(event string, handler EventHandler) {
	b.handlers[event] = append(b.handlers[event], handler)
}

// Emit dispatches synchronously; handler panics are logged, not propagated.
func (b *EventBus) Emit(event string, payload map[string]any) {
	record := map[string]any{}
	for k, v := range payload {
		record[k] = v
	}
	b.History = append(b.History, EventRecord{Event: event, Payload: record})
	for _, handler := range b.handlers[event] {
		func() {
			defer func() {
				if r := recover(); r != nil {
					log.Printf("event_handler_failed event=%s err=%v", event, r)
				}
			}()
			handler(record)
		}()
	}
}
